crates/ln-dlc-node/src/channel.rs

```diff
@@ -0,0 +1,189 @@
+    pub counterparty: PublicKey,
+    pub created_at: OffsetDateTime,
+    pub updated_at: OffsetDateTime,
+}
+
+impl Channel {
+    pub fn new(inbound: u64, outbound: u64, counterparty: PublicKey) -> Self {
+        Channel {
+            user_channel_id: UserChannelId::new(),
+            channel_state: ChannelState::Pending,
+            inbound,
+            outbound,
+            counterparty,
+            created_at: OffsetDateTime::now_utc(),
+            updated_at: OffsetDateTime::now_utc(),
+            channel_id: None,
+            funding_txid: None,
+        }
+    }
+
+    pub fn is_closed(&self) -> bool {
+        matches!(
+            self.channel_state,
+            ChannelState::ForceClosedLocal | ChannelState::ForceClosedRemote | ChannelState::Closed
+        )
+    }
+
+    pub fn close_channel(channel: Channel, reason: ClosureReason) -> Channel {
+        if channel.is_closed() {
+            tracing::warn!(%channel, "Unexpected state transition. Expected channel state to be either 'Pending' or 'Open', but was '{:?}'", channel.channel_state);
+        }
+
+        let mut channel = channel;
+        channel.channel_state = reason.into();
+        channel.updated_at = OffsetDateTime::now_utc();
+        channel
+    }
+
+    pub fn update_liquidty(
+        channel: &Channel,
+        channel_details: &Option<ChannelDetails>,
+    ) -> Result<Channel> {
+        let mut channel = channel.clone();
+        match channel_details {
+            Some(channel_details) => {
+                channel.inbound = channel_details.inbound_capacity_msat / 1000;
+                channel.outbound = channel_details.outbound_capacity_msat / 1000;
+                Ok(channel)
+            }
+            None => {
+                bail!("Couldn't find channel details");
+            }
+        }
+    }
+
+    pub fn open_channel(
+        channel: Option<Channel>,
+        channel_details: ChannelDetails,
+    ) -> Result<Channel> {
+        let mut channel = match channel {
+            Some(channel) => channel,
+            None => {
+                let user_channel_id =
+                    UserChannelId::from(channel_details.user_channel_id).to_string();
+                tracing::warn!(%user_channel_id, channel_id = %channel_details.channel_id.to_hex(), public = channel_details.is_public, outbound = channel_details.is_outbound, "Creating a new shadow channel");
+                Channel::new(
+                    channel_details.inbound_capacity_msat / 1000,
+                    0,
+                    channel_details.counterparty.node_id,
+                )
+            }
+        };
+
+        tracing::debug!("Updating shadow channel.");
+        channel.channel_state = ChannelState::Open;
+        channel.funding_txid = channel_details.funding_txo.map(|txo| txo.txid);
+        channel.channel_id = Some(channel_details.channel_id);
+        channel.updated_at = OffsetDateTime::now_utc();
+        Ok(channel)
```

Comment:
🔧 Probably should log the whole channel after mutating!

```suggestion
        channel.channel_state = ChannelState::Open;
        channel.funding_txid = channel_details.funding_txo.map(|txo| txo.txid);
        channel.channel_id = Some(channel_details.channel_id);
        channel.updated_at = OffsetDateTime::now_utc();

        tracing::debug!(%channel, "Set shadow channel to open");

        Ok(channel)
```